### PR TITLE
feat: add user_policy_document parameter

### PR DESCRIPTION
**Pull request by dod38fr**
## what

This parameter allows the user to specify policies that are applied to the S3 bucket with the policies defined by this module.

## why

We want to add policies that forbid non admin users to access the bucket containing tfstates.

This commit allow us to specify a policy that implement these restriction without clobbering the policies put in place by this module.

Note that I have no problem to change the name of this new parameter if you want another.

## references

Closes: #115

### CODE REVIEW

**Comment by dod38fr** [comment]
Hi
Any news ?

**Comment by dod38fr** [comment]
Hi

Any news about this PR ? Is there a problem preventing a review ?

All the best

**Comment by osterman** [comment]
Best way to get attention is in #pr-reviews channel in slack

File: variables.tf [hunk 1]
```diff
@@ -189,3 +189,9 @@ variable "permissions_boundary" {
   default     = ""
   description = "ARN of the policy that is used to set the permissions boundary for the IAM replication role"
 }
+
+variable "user_policy_documents" {
+  type        = list(string)
+  default     = []
+  description = "List of IAM policy documents (in JSON format) to attach to the S3 bucket in addition of the policies created by this module."
+}
```
Comment (Nuru) [inline variables.tf:197]:
```suggestion
variable "source_policy_documents" {
  type        = list(string)
  default     = []
  description = <<-EOT
    List of IAM policy documents (in JSON format) that are merged together into the exported document.
    Statements must have unique SIDs.
    Statement having SIDs that match policy SIDs generated by this module will override them.
    EOT
}
```

File: main.tf [hunk 2]
```diff
@@ -56,6 +56,8 @@ data "aws_region" "current" {}
 data "aws_iam_policy_document" "bucket_policy" {
   count = local.enabled ? 1 : 0
 
+  source_policy_documents = var.user_policy_documents
+
```
Comment (Nuru) [inline main.tf:60]: Move to separate policy document which overrides the default bucket policy. See s3-bucket implementation for example:

https://github.com/cloudposse/terraform-aws-s3-bucket/blob/d7a49439ea48dfd91c06830fb5d9419329bd927d/main.tf#L454-L466

**Review by Nuru — CHANGES_REQUESTED** [review]
Sorry for the delay reviewing this. We were hoping to make these changes and more by now as part of a overhaul, but obviously we have not gotten to it yet. 

Please make the requested changes so that this feature works just like `source_policy_documents` in [terraform-aws-s3-bucket](https://github.com/cloudposse/terraform-aws-s3-bucket).

File: main.tf (hunk 2 above)
Comment (dod38fr) [reply main.tf]: Done.

Unfortunately, I won't be able to test this change on my client's system before October.

**Comment by dod38fr** [comment]
> please run the following commands and commit the changes:
> 
> ```shell
> make init
> make github/init
> make readme
> ```

Done.

All the best

File: main.tf (hunk 2 above)
Comment (vtavernier) [reply main.tf]: I can confirm that the change works as expected on said client's system where we are running the changed version in this PR, thanks @dod38fr!

Comment (dod38fr) [reply main.tf]: All set then, @Nuru , can you resume the review ?

All the best

**Comment by dod38fr** [comment]
@Nuru The terratest job (https://github.com/cloudposse/terraform-aws-tfstate-backend/actions/runs/6303986183) is now marked as successful.

I don't really know what went on. I guess that he failure you've seen was a temporary failure and that the job was restarted.

In any case, I cannot investigate further as I cannot see any error message.

Is there something more I can do ?

**Comment by dod38fr** [comment]
Hi

I've successfully run the test that failed in your CI:
```
$ make -C test/src/
make: Entering directory '/home/domi/terraform-stuff/terraform-aws-tfstate-backend/test/src'
go mod download
go test -v -timeout 60m
=== RUN   TestExamplesComplete
=== PAUSE TestExamplesComplete
=== RUN   TestExamplesCompleteDisabled
=== PAUSE TestExamplesCompleteDisabled
=== CONT  TestExamplesComplete
=== CONT  TestExamplesCompleteDisabled
TestExamplesComplete 2023-10-18T17:53:30+02:00 test_structure.go:130: Copied terraform folder ../../examples/complete to /tmp/TestExamplesComplete578001989/terraform-aws-tfstate-backend/examples/complete
TestExamplesComplete 2023-10-18T17:53:30+02:00 retry.go:91: terraform [init -upgrade=true]
TestExamplesComplete 2023-10-18T17:53:30+02:00 logger.go:66: Running command terraform with args [init -upgrade=true]

[ snip]

TestExamplesComplete 2023-10-18T17:59:00+02:00 logger.go:66: module.tfstate_backend.aws_s3_bucket.default[0]: Still destroying... [id=eg-use2-test-terraform-tfstate-backend-rxcfw2, 4m40s elapsed]
TestExamplesComplete 2023-10-18T17:59:10+02:00 logger.go:66: module.tfstate_backend.aws_s3_bucket.default[0]: Still destroying... [id=eg-use2-test-terraform-tfstate-backend-rxcfw2, 4m50s elapsed]
TestExamplesComplete 2023-10-18T17:59:20+02:00 logger.go:66: module.tfstate_backend.aws_s3_bucket.default[0]: Still destroying... [id=eg-use2-test-terraform-tfstate-backend-rxcfw2, 5m0s elapsed]
TestExamplesComplete 2023-10-18T17:59:21+02:00 logger.go:66: module.tfstate_backend.aws_s3_bucket.default[0]: Destruction complete after 5m2s
TestExamplesComplete 2023-10-18T17:59:21+02:00 logger.go:66: 
TestExamplesComplete 2023-10-18T17:59:21+02:00 logger.go:66: Destroy complete! Resources: 9 destroyed.
TestExamplesComplete 2023-10-18T17:59:21+02:00 logger.go:66: 
--- PASS: TestExamplesComplete (351.46s)
PASS
ok      github.com/cloudposse/terraform-aws-tfstate-backend     351.483s
make: Leaving directory '/home/domi/terraform-stuff/terraform-aws-tfstate-backend/test/src'
```

I really don't know what can be done to unblock this PR.

Any idea ?

**Review by Nuru — APPROVED** [review]

**Pull request merged**